### PR TITLE
Feature reset password

### PR DESCRIPTION
**Pull request by singhsanket143**
The pull request is implementation for the feature listed in issue #16615. The forgot password functionality has been added for already registered users with proper error message displays and email for reset password link. After successful password reset the user is automatically logged in. Invalid emails or nonmatching passwords gives proper error notice.

1. Added reset password link in the login modal
image
2. Forgot password email
image
3. Added email for reset password instructions
image
4. Error message for unregistered email
image
5. Reset password modal with reset password token in the URL bar
image
6. Error for passwords donot match
image
7. Login after successful password reset
image

### CODE REVIEW

File: client/styles/global.scss [hunk 1]
```diff
@@ -133,6 +133,47 @@ a {
...
+	color: white;
+	padding-top: 5px;
+
+  &:hover {
+		background-color: #4cd785;
+	}
+}
+
+#resetpasswordsubmitbutton {
+  width: 40%;
```
Comment (peter-hank) [inline client/styles/global.scss]: indentation

File: client/views/resetPassword/resetPassword.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  const parentNode = document.getElementsByClassName(parentElement)[0];
+  const nextNode = document.getElementById(nextElement);
+  currentError = Blaze.renderWithData(Template.form_error, reason, parentNode, nextNode);
+}
+
+Template.ResetPassword.events({
+  'click #resetpasswordsubmitbutton': function(event, template) {
+    event.preventDefault();
+    var password = $("#resetPassword").val(); // Get the new password
```
Comment (peter-hank) [inline client/views/resetPassword/resetPassword.js]: comments above or below the code please

Comment (peter-hank) [reply client/views/resetPassword/resetPassword.js]: single quotes as well

File: client/views/resetPassword/resetPassword.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    else if(password.length < 5) { // If length of password is less than 6
+      showError('Password is not strong enough', 'inputcontainer', 'resetPassword');
+    }
+    else { // Update the password
+      Accounts.resetPassword(Router.current().params.token, password, function(err){
+        if(err) { // If some internal server error occurs
+          console.log('Something went wrong');
+          showError('Something went wrong', 'inputcontainer', 'resetPassword');
+        } else { 
+          console.log('Password changed');
```
Comment (peter-hank) [inline client/views/resetPassword/resetPassword.js]: `Password changed` > `Password has been changed`

File: client/views/forgotPassword/forgotPassword.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  var reg = /^([A-Za-z0-9_\-\.])+\@([A-Za-z0-9_\-\.])+\.([A-Za-z]{2,4})$/;
+  if(reg.test(email) === false) {
+    console.log('Invalid email');
+    return false;
+  }
+  return true;
+}
+
+function showError(reason, parentElement, nextElement) {
+  if (typeof currentError !== 'undefined') {
```
Comment (peter-hank) [inline client/views/forgotPassword/forgotPassword.js:12]: let's be consistent and put a space after `if` everywhere

Comment (peter-hank) [inline client/views/forgotPassword/forgotPassword.js:12]: let's be consistent and put a space after `if` everywhere

**Review by peter-hank — CHANGES_REQUESTED** [review]
Ok, we need some updates here.

File: client/views/forgotPassword/forgotPassword.js (hunk 4 above)
Comment (singhsanket143) [reply client/views/forgotPassword/forgotPassword.js:12]: @peter-hank Add spaces everywhere after if conditional

Comment (singhsanket143) [reply client/views/forgotPassword/forgotPassword.js:12]: @peter-hank Add spaces everywhere after if conditional

File: client/views/resetPassword/resetPassword.js (hunk 3 above)
Comment (singhsanket143) [reply client/views/resetPassword/resetPassword.js]: @peter-hank Updated the log message

File: client/views/resetPassword/resetPassword.js (hunk 2 above)
Comment (singhsanket143) [reply client/views/resetPassword/resetPassword.js]: @peter-hank Updated the quotes style

File: client/styles/global.scss (hunk 1 above)
Comment (singhsanket143) [reply client/styles/global.scss]: @peter-hank I have made some changes for indentation. Please review them and if there are any more suggestions then do let me know

Comment (peter-hank) [reply client/styles/global.scss]: @singhsanket143 can you use some online tool like http://www.calluna-software.com/Formatter/Css and fix the whole file?

Comment (singhsanket143) [reply client/styles/global.scss]: @peter-hank Updated the indentation

Comment (peter-hank) [reply client/styles/global.scss]: @singhsanket143 it's still not nice, can you make sure the whole file looks good?

Comment (singhsanket143) [reply client/styles/global.scss]: @peter-hank I have used https://www.freeformatter.com/css-beautifier.html as a tool to indent the whole file. Please review it.

File: client/views/login/login.html [hunk 5]
```diff
@@ -13,9 +13,12 @@ <h1 class="formtitle">Login</h1>
 			<div class="modalbottommessage">
 				Don't have an account? <p class="modalemphasis" id="registeremphasis">Register</p>
 			</div>
+			<div class="modalbottommessage">
+				Forgot password? <p class="modalemphasis" id="forgotpasswordemphasis">Send instruction</p>
```
Comment (peter-hank) [inline client/views/login/login.html]: @singhsanket143 `Send instruction` > `Send instructions`

File: client/views/forgotPassword/forgotPassword.html [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<template name="ForgotPassword">
+	<div id="darker">
+	</div>
+	<div class="formcontainer">
+		{{> close_button}}
+		<h1 class="formtitle">Forgot Password</h1>
+		<div class="modaltopmessage">
+			Enter your registered Email
```
Comment (peter-hank) [inline client/views/forgotPassword/forgotPassword.html]: @singhsanket143 `Enter your registered Email` > `Enter your registered email address`

Comment (singhsanket143) [reply client/views/forgotPassword/forgotPassword.html]: @peter-hank I have updated the text here

File: client/views/login/login.html (hunk 5 above)
Comment (singhsanket143) [reply client/views/login/login.html]: @peter-hank I have updated the text here

File: client/views/forgotPassword/forgotPassword.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        console.log(err)
+        if (err.message === 'User not found [403]') {
+          showError('Email doesn\'t exist', 'inputcontainer', 'forgotPasswordEmail');
+          console.log('Email doesn\'t exist');
+        } else {
+          showError('We are sorry, but something went wrong', 'inputcontainer', 'forgotPasswordEmail');
+          console.log('We are sorry, but something went wrong');
+        }
+      } else {
+        console.log('Email sent, check your mailbox');
```
Comment (peter-hank) [inline client/views/forgotPassword/forgotPassword.js]: Can we expose this message to the user?

**Comment by peter-hank** [comment]
@singhsanket143 

We also need to make it possible to configure a mail server using env variables, currently it's hardcoded to use localhost.

File: client/views/forgotPassword/forgotPassword.js (hunk 7 above)
Comment (singhsanket143) [reply client/views/forgotPassword/forgotPassword.js]: @peter-hank Yes, we can add a popup notification using some library. Shall I Do that?

**Comment by singhsanket143** [comment]
> @singhsanket143
> 
> We also need to make it possible to configure a mail server using env variables, currently it's hardcoded to use localhost.

@peter-hank So this thing will be configured on the production server right?

**Comment by peter-hank** [comment]
> > @singhsanket143
> > We also need to make it possible to configure a mail server using env variables, currently it's hardcoded to use localhost.
> 
> @peter-hank So this thing will be configured on the production server right?

In development as well.

File: client/views/forgotPassword/forgotPassword.js (hunk 7 above)
Comment (peter-hank) [reply client/views/forgotPassword/forgotPassword.js]: > @peter-hank Yes, we can add a popup notification using some library. Shall I Do that?

There's nothing there already?

**Comment by singhsanket143** [comment]
> > > @singhsanket143
> > > We also need to make it possible to configure a mail server using env variables, currently it's hardcoded to use localhost.
> > 
> > 
> > @peter-hank So this thing will be configured on the production server right?
> 
> In development as well.

@peter-hank So shall we do this in this pull request only?

**Comment by singhsanket143** [comment]
> > > > @singhsanket143
> > > > We also need to make it possible to configure a mail server using env variables, currently it's hardcoded to use localhost.
> > > 
> > > 
> > > @peter-hank So this thing will be configured on the production server right?
> > 
> > 
> > In development as well.
> 
> @peter-hank So shall we do this in this pull request only?

@peter-hank I have updated this

File: client/views/forgotPassword/forgotPassword.js (hunk 7 above)
Comment (peter-hank) [inline client/views/forgotPassword/forgotPassword.js]: We should expose this message to the user as well.

Comment (peter-hank) [reply client/views/forgotPassword/forgotPassword.js]: @singhsanket143

**Comment by peter-hank** [comment]
:rainbow:

**Pull request merged**